cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/output/MultiOutputCommitter.java

```diff
@@ -122,8 +112,67 @@ public boolean isRecoverySupported() {
 
   @Override
   public void recoverTask(TaskAttemptContext taskContext) throws IOException {
+    delegateInParallel((name, delegate) -> {
+      TaskAttemptContext namedContext = MultiOutputFormat.getNamedTaskContext(taskContext, name);
+      delegate.recoverTask(namedContext);
+    });
+  }
+
+  private void delegateInParallel(DelegateFunction delegateFunction) throws IOException {
+    ExecutorService executorService = Executors.newFixedThreadPool(delegates.size(), new ThreadFactory() {
+      AtomicInteger id = new AtomicInteger(0);
+
+      @Override
+      public Thread newThread(Runnable r) {
+        Thread t = new Thread(r);
+        t.setName(String.format("multi-output-committer-%d", id.getAndIncrement()));
+        return t;
+      }
+    });
+
+    List<Future<Void>> futures = new ArrayList<>(delegates.size());
     for (Map.Entry<String, OutputCommitter> entry : delegates.entrySet()) {
```

Comment:
The code can be simplified by using `CompletionService` (`ExecutorCompletionService`)